Core/include/Acts/TrackFitting/MbfSmoother.hpp

```diff
@@ -0,0 +1,138 @@
+// This file is part of the Acts project.
+//
+// Copyright (C) 2024 CERN for the benefit of the Acts project
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+#pragma once
+
+#include "Acts/Definitions/TrackParametrization.hpp"
+#include "Acts/EventData/MultiTrajectory.hpp"
+#include "Acts/Geometry/GeometryContext.hpp"
+#include "Acts/TrackFitting/KalmanFitterError.hpp"
+#include "Acts/Utilities/Delegate.hpp"
+#include "Acts/Utilities/Logger.hpp"
+#include "Acts/Utilities/Result.hpp"
+
+#include <cassert>
+#include <cstddef>
+#include <system_error>
+
+namespace Acts {
+
+/// Kalman trajectory smoother based on the Modified Bryson–Frazier smoother.
+///
+/// This implements not a single smoothing step, but the full backwards
+/// smoothing procedure for a filtered, forward trajectory using the stored
+/// linearization.
+class MbfSmoother {
+ public:
+  struct InternalTrackState {
+    using Projector =
+        typename TrackStateTraits<MultiTrajectoryTraits::MeasurementSizeMax,
+                                  false>::Projector;
+    using Jacobian =
+        typename TrackStateTraits<MultiTrajectoryTraits::MeasurementSizeMax,
+                                  false>::Covariance;
+    using Parameters =
+        typename TrackStateTraits<MultiTrajectoryTraits::MeasurementSizeMax,
+                                  false>::Parameters;
+    using Covariance =
+        typename TrackStateTraits<MultiTrajectoryTraits::MeasurementSizeMax,
+                                  false>::Covariance;
+
+    // This is used to build a covariance matrix view in the .cpp file
+    unsigned int calibratedSize{0};
+    const double* calibrated{nullptr};
+    const double* calibratedCovariance{nullptr};
+    Projector projector;
```

Comment:
Maybe initialize to projector{nullptr}, similar to the other members, i.e. Jacobian/parameters/covariance?